wallet/wallet.go

```diff
@@ -685,108 +678,136 @@ func (w *Wallet) recovery(startHeight int32) error {
-			err := w.recoverDefaultScopes(
-				chainClient, tx, addrMgrNS, blockBatch,
-				recoveryMgr.State(),
-			)
+	for height < bestHeight {
+		// We'll perform our recovery in batches of 2000 blocks. It's
+		// possible for us to run into blocks before our birthday if our
+		// birthday is after our reorg safe height, so we'll make sure
+		// to not add those to the batch.
+		for len(recoveryMgr.BlockBatch()) < recoveryBatchSize {
+			hash, err := chainClient.GetBlockHash(int64(height))
 			if err != nil {
 				return err
 			}
-
-			// Clear the batch of all processed blocks.
-			recoveryMgr.ResetBlockBatch()
-
-			if err := tx.Commit(); err != nil {
+			header, err := chainClient.GetBlockHeader(hash)
+			if err != nil {
 				return err
 			}
+			blocks = append(blocks, &waddrmgr.BlockStamp{
+				Hash:      *hash,
+				Height:    height,
+				Timestamp: header.Timestamp,
+			})
 
-			log.Infof("Recovered addresses from blocks %d-%d",
-				blockBatch[0].Height,
-				blockBatch[len(blockBatch)-1].Height)
+			if height >= birthdayBlock.Height {
+				recoveryMgr.AddToBlockBatch(
+					hash, height, header.Timestamp,
+				)
+			}
 
-			tx, err = w.db.BeginReadWriteTx()
-			if err != nil {
-				return err
+			// It's possible for us to reach our best height without
+			// exceeding the recovery batch size, so we can proceed
+			// to commit our state to disk.
+			if height == bestHeight {
+				break
 			}
-			addrMgrNS = tx.ReadWriteBucket(waddrmgrNamespaceKey)
+			height++
 		}
 
-		// Since the recovery in a way acts as a rescan, we'll update
-		// the wallet's tip to point to the current block so that we
-		// don't unnecessarily rescan the same block again later on.
-		return w.Manager.SetSyncedTo(addrMgrNS, &waddrmgr.BlockStamp{
-			Hash:      *hash,
-			Height:    height,
-			Timestamp: header.Timestamp,
+		batch := recoveryMgr.BlockBatch()
+		err := walletdb.Update(w.db, func(tx walletdb.ReadWriteTx) error {
+			ns := tx.ReadWriteBucket(waddrmgrNamespaceKey)
+			for _, block := range blocks {
+				err := w.Manager.SetSyncedTo(ns, block)
+				if err != nil {
+					return err
+				}
+			}
+			return w.recoverDefaultScopes(
+				chainClient, tx, ns, batch, recoveryMgr.State(),
+			)
 		})
-	})
-	if err != nil {
-		tx.Rollback()
-		return err
+		if err != nil {
+			return err
+		}
+
+		if len(batch) > 0 {
+			log.Infof("Recovered addresses from blocks %d-%d",
+				batch[0].Height, batch[len(batch)-1].Height)
+		}
+
+		// Clear the batch of all processed blocks to reuse the same
+		// memory for future batches.
+		blocks = blocks[:0]
+		recoveryMgr.ResetBlockBatch()
 	}
 
-	// Now that we've reached the chain tip, we can process our final batch
-	// with the remaining blocks if it did not reach its maximum size.
-	blockBatch := recoveryMgr.BlockBatch()
-	err = w.recoverDefaultScopes(
-		chainClient, tx, addrMgrNS, blockBatch, recoveryMgr.State(),
-	)
+	return nil
+}
+
+// syncHeightRange determines the best height range to sync with the chain to
```

Comment:
naming suggestion: `getSyncRange`? Since current name indicates something is being synced when this method is called :P

Reply:
Done.